[Feature Request]Handle NoSuchUploadException in UploadPart operation to prevent broker crash

## Title

Handle NoSuchUploadException in UploadPart operation to prevent broker crash

## Description

When using AutoMQ with Tencent Cloud COS (or other S3-compatible storage), the broker crashes with `Runtime.getRuntime().halt(1)` when encountering `NoSuchUploadException` during `UploadPart` operations.

### Root Cause

The current implementation in `AwsObjectStorage.toRetryStrategyAndCause()` only handles `NoSuchUploadException` specially for `COMPLETE_MULTI_PART_UPLOAD` operation:

```java
if (COMPLETE_MULTI_PART_UPLOAD == operation) {
    if (cause instanceof NoSuchUploadException) {
        strategy = RetryStrategy.VISIBILITY_CHECK;
    }
}
```

However, for `UPLOAD_PART` operation, `NoSuchUploadException` (HTTP 404) results in `RetryStrategy.ABORT`, which eventually propagates to `S3Storage.commitDeltaWALUpload()` and triggers:

```java
Runtime.getRuntime().halt(1);
```

### Scenario

This issue occurs when:
1. A multipart upload is initiated
2. The upload takes longer than expected (due to network issues, throttling, or large data)
3. The cloud storage's lifecycle rule automatically aborts incomplete multipart uploads (e.g., after 1-7 days)
4. Subsequent `UploadPart` calls fail with `NoSuchUploadException`
5. Broker crashes

### Error Log

```
[ERROR] UploadPart for object 2b180480/_kafka_ops_sh/138445234-2 fail (com.automq.stream.s3.operator.AbstractObjectStorage)
software.amazon.awssdk.services.s3.model.NoSuchUploadException: The specified multipart upload does not exist.
The upload ID might be invalid, or the multipart upload might have been aborted or completed.

[ERROR] Unexpected exception when commit stream set object (com.automq.stream.s3.S3Storage)
java.util.concurrent.CompletionException: software.amazon.awssdk.services.s3.model.NoSuchUploadException: ...
```

### Proposed Solution

1. When `NoSuchUploadException` occurs during `UPLOAD_PART`, instead of aborting, the system should:
   - Invalidate the current uploadId
   - Re-initiate a new multipart upload
   - Retry the upload from the beginning

2. Add uploadId validity tracking to detect stale upload sessions early

3. Consider adding a configurable timeout for multipart uploads to proactively restart uploads that are taking too long

### Environment

- Cloud Provider: Tencent Cloud COS (S3-compatible)
- The bucket has lifecycle rules configured to abort incomplete multipart uploads

### Impact

- Severity: High
- Broker crashes and requires manual restart
- Data durability may be affected if WAL upload fails

--- a/s3stream/src/main/java/com/automq/stream/s3/operator/AwsObjectStorage.java
+++ b/s3stream/src/main/java/com/automq/stream/s3/operator/AwsObjectStorage.java
@@ -415,9 +415,13 @@ public class AwsObjectStorage extends AbstractObjectStorage {
                 default:
                     strategy = RetryStrategy.RETRY;
             }
-            if (COMPLETE_MULTI_PART_UPLOAD == operation) {
-                if (cause instanceof NoSuchUploadException) {
+            if (cause instanceof NoSuchUploadException) {
+                if (COMPLETE_MULTI_PART_UPLOAD == operation) {
                     strategy = RetryStrategy.VISIBILITY_CHECK;
+                } else if (S3Operation.UPLOAD_PART == operation || S3Operation.UPLOAD_PART_COPY == operation) {
+                    // The multipart upload no longer exists (expired, aborted, or cleaned by the backend).
+                    // Retrying with a dead upload ID is futile and would loop until the broker crashes.
+                    strategy = RetryStrategy.ABORT;
                 }
             }
             if (GET_OBJECT == operation) {

--- a/s3stream/src/test/java/com/automq/stream/s3/operator/AwsObjectStorageTest.java
+++ b/s3stream/src/test/java/com/automq/stream/s3/operator/AwsObjectStorageTest.java
@@ -1,6 +1,9 @@
 package com.automq.stream.s3.operator;
 
 import com.automq.stream.s3.TestUtils;
+import com.automq.stream.s3.metrics.operations.S3Operation;
+
+import org.apache.commons.lang3.tuple.Pair;
 
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
@@ -24,6 +27,7 @@ import software.amazon.awssdk.core.async.AsyncRequestBody;
 import software.amazon.awssdk.services.s3.S3AsyncClient;
 import software.amazon.awssdk.services.s3.model.ChecksumAlgorithm;
 import software.amazon.awssdk.services.s3.model.DeleteObjectsResponse;
+import software.amazon.awssdk.services.s3.model.NoSuchUploadException;
 import software.amazon.awssdk.services.s3.model.PutObjectRequest;
 import software.amazon.awssdk.services.s3.model.S3Error;
 
@@ -142,6 +146,62 @@ public class AwsObjectStorageTest {
         ByteBuffer value = ByteBuffer.allocate(Integer.BYTES);
         value.putInt((int) crc32c.getValue());
         return Base64.getEncoder().encodeToString(value.array());
+    }
+
+
+    /**
+     * Given a NoSuchUploadException during UPLOAD_PART,
+     * When toRetryStrategyAndCause is called,
+     * Then the strategy should be ABORT because retrying with a dead upload ID is futile.
+     */
+    @Test
+    void testNoSuchUploadExceptionAbortsUploadPart() {
+        AwsObjectStorage storage = new AwsObjectStorage(null, "bucket");
+        NoSuchUploadException cause = NoSuchUploadException.builder()
+            .message("The specified upload does not exist")
+            .statusCode(404)
+            .build();
+
+        Pair<RetryStrategy, Throwable> result = storage.toRetryStrategyAndCause(cause, S3Operation.UPLOAD_PART);
+        Assertions.assertEquals(RetryStrategy.ABORT, result.getLeft(),
+            "UPLOAD_PART with NoSuchUploadException should ABORT, not retry indefinitely");
+    }
+
+    /**
+     * Given a NoSuchUploadException during UPLOAD_PART_COPY,
+     * When toRetryStrategyAndCause is called,
+     * Then the strategy should be ABORT for the same reason as UPLOAD_PART.
+     */
+    @Test
+    void testNoSuchUploadExceptionAbortsUploadPartCopy() {
+        AwsObjectStorage storage = new AwsObjectStorage(null, "bucket");
+        NoSuchUploadException cause = NoSuchUploadException.builder()
+            .message("The specified upload does not exist")
+            .statusCode(404)
+            .build();
+
+        Pair<RetryStrategy, Throwable> result = storage.toRetryStrategyAndCause(cause, S3Operation.UPLOAD_PART_COPY);
+        Assertions.assertEquals(RetryStrategy.ABORT, result.getLeft(),
+            "UPLOAD_PART_COPY with NoSuchUploadException should ABORT, not retry indefinitely");
+    }
+
+    /**
+     * Given a NoSuchUploadException during COMPLETE_MULTI_PART_UPLOAD,
+     * When toRetryStrategyAndCause is called,
+     * Then the strategy should be VISIBILITY_CHECK because the upload may have already completed
+     * and the object may just not be visible yet.
+     */
+    @Test
+    void testNoSuchUploadExceptionVisibilityCheckForComplete() {
+        AwsObjectStorage storage = new AwsObjectStorage(null, "bucket");
+        NoSuchUploadException cause = NoSuchUploadException.builder()
+            .message("The specified upload does not exist")
+            .statusCode(404)
+            .build();
+
+        Pair<RetryStrategy, Throwable> result = storage.toRetryStrategyAndCause(cause, S3Operation.COMPLETE_MULTI_PART_UPLOAD);
+        Assertions.assertEquals(RetryStrategy.VISIBILITY_CHECK, result.getLeft(),
+            "COMPLETE_MULTI_PART_UPLOAD with NoSuchUploadException should VISIBILITY_CHECK");
     }
 
 }

--- a/s3stream/src/test/java/com/automq/stream/s3/operator/AwsObjectStorageTest.java
+++ b/s3stream/src/test/java/com/automq/stream/s3/operator/AwsObjectStorageTest.java
@@ -4,7 +4,6 @@ import com.automq.stream.s3.TestUtils;
 import com.automq.stream.s3.metrics.operations.S3Operation;
 
 import org.apache.commons.lang3.tuple.Pair;
-
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 

